### PR TITLE
Document the new OpenID Connect feature for Crowbar (SOC-10510)

### PR DESCRIPTION
**Pull request by guangyee**
Document how to enable and configure OpenID Connect feature
in Crowbar.

### CODE REVIEW

File: xml/depl_nodes.xml [hunk 1]
```diff
@@ -1635,6 +1635,209 @@ and admins can see the available options in the barclamp -->
...
   </sect2>
+
+  <sect2 xml:id="sec-depl-ostack-keystone-openidc">
+   <title>OpenID Connect Setup for &o_ident;</title>
+   <para>
+    &o_ident; supports WebSSO by federating with an external identity
+    provider (IdP) using  <link xlink:href="https://github.com/zmartzone/mod_auth_openidc">auth_openidc</link> module.
+   </para>
+   <para>
+    There are two steps to enable this feature.
```
Comment (asettle) [inline xml/depl_nodes.xml]:
```suggestion
There are two steps to enable this feature:

File: xml/depl_nodes.xml [hunk 2]
```diff
@@ -1635,6 +1635,209 @@ and admins can see the available options in the barclamp -->
...
+   </para>
+   <orderedlist>
+     <listitem>
+      <para>
+       Configure the "federation" and "openidc" attributes for the Keystone Barclamp in Crowbar.
```
Comment (asettle) [inline xml/depl_nodes.xml]:
```suggestion
Configure the <literal>federation</literal> and  <literal>openidc</literal> attributes for the Keystone Barclamp in Crowbar.

File: xml/depl_nodes.xml [hunk 3]
```diff
@@ -1635,6 +1635,209 @@ and admins can see the available options in the barclamp -->
...
+       &o_ident; using <guimenu>OpenStack Command Line Tool (CLI)</guimenu>.
+      </para>
+     </listitem>
+   </orderedlist>
+   <sect3 xml:id="sec-depl-ostack-keystone-openidc-crowbar">
+   <title>&o_ident; Barclamp Configuration</title>
+   <para>
+    Configurating OpenID Connect is done in the <guimenu>Raw</guimenu> view,
+    under the <guimenu>"federation"</guimenu> section. The global attributes,
+    namely <guimenu>"trusted_dashboards"</guimenu> and <guimenu>"websso_keystone_url"</guimenu>
```
Comment (asettle) [inline xml/depl_nodes.xml]: ditto here

File: xml/depl_nodes.xml [hunk 4]
```diff
@@ -1635,6 +1635,209 @@ and admins can see the available options in the barclamp -->
...
+    are not specific to OpenID Connect. Rather, they are designed to help
+    facilitate WebSSO browser redirects with external IdPs in a complex cloud
+    deployment environment.
+   </para>
+   <para>
+    If the cloud deployment does not have any external proxies or load
+    balancers, where the public Keystone and Horizon (a.k.a. Dashboard)
```
Comment (asettle) [inline xml/depl_nodes.xml]:
```suggestion
balancers, where the public keystone and horizon (Dashboard)

File: xml/depl_nodes.xml (hunk 1 above)
Comment (guangyee) [reply xml/depl_nodes.xml]: Done

File: xml/depl_nodes.xml (hunk 3 above)
Comment (guangyee) [reply xml/depl_nodes.xml]: Done

File: xml/depl_nodes.xml [hunk 5]
```diff
@@ -1635,6 +1635,209 @@ and admins can see the available options in the barclamp -->
...
+       Create the Identity Provider, Protocol, and Mapping resource in
+       &o_ident; using <guimenu>OpenStack Command Line Tool (CLI)</guimenu>.
+      </para>
+     </listitem>
+   </orderedlist>
+   <sect3 xml:id="sec-depl-ostack-keystone-openidc-crowbar">
+   <title>&o_ident; Barclamp Configuration</title>
+   <para>
+    Configurating OpenID Connect is done in the <guimenu>Raw</guimenu> view,
+    under the <guimenu>"federation"</guimenu> section. The global attributes,
```
[the remark this answers is not in the document]
Comment (guangyee) [reply xml/depl_nodes.xml]: Done

File: xml/depl_nodes.xml (hunk 4 above)
Comment (guangyee) [reply xml/depl_nodes.xml]: Done

File: xml/depl_nodes.xml [hunk 6]
```diff
@@ -1635,6 +1635,209 @@ and admins can see the available options in the barclamp -->
...
+             },
+             {
+                 "type": "HTTP_OIDC_EMAIL"
+             }
+
+        ]
+    }
+]</screen>
+     <para>
+      Once the mapping file is created, now create the Mapping resource in Keystone. For example:
```
Comment (asettle) [inline xml/depl_nodes.xml]:
```suggestion
      Once the mapping file is created, now create the mapping resource in Keystone. For example:

File: xml/depl_nodes.xml [hunk 7]
```diff
@@ -1635,6 +1635,209 @@ and admins can see the available options in the barclamp -->
...
+    endpoints are directly managed by Crowbar, <guimenu>trusted_dashboards</guimenu>
+    and <guimenu>websso_keystone_url</guimenu> does not need to be provided.
+    However, in a complex cloud deployment where the public Keystone and
+    Horizon endpoints are handled by external load balancers or proxies, and
+    they are not directly managed by Crowbar, <guimenu>trusted_dashboards</guimenu>
+    and <guimenu>websso_keystone_url</guimenu> must be provided and they must
+    correctly reflect the external public endpoints.
+   </para>
+   <para>
+    To configure OpenID Connect, edit the attributes in the <guimenu>"openidc"</guimenu> subsection.
```
[the remark this answers is not in the document]
Comment (guangyee) [reply xml/depl_nodes.xml]: Done

File: xml/depl_nodes.xml [hunk 8]
```diff
@@ -1635,6 +1635,209 @@ and admins can see the available options in the barclamp -->
...
+   </para>
+   <procedure>
+    <step>
+     <para>
+      Set <guimenu>"enabled":true</guimenu>
```
[the remark this answers is not in the document]
Comment (guangyee) [reply xml/depl_nodes.xml]: But those are what it appears in the raw view though. The strings in the raw view are in quotes. See the Keystone LDAP example here.

http://docserv.nue.suse.com/documents/SUSE_OpenStack_Cloud_9/suse-openstack-cloud-crowbar-deployment/single-html/#sec-depl-ostack-keystone-ldap

[the remark this answers is not in the document]
Comment (asettle) [reply xml/depl_nodes.xml]: That is *weird*. Is that how it appears in the GUI?

**Pull request merged**